iwyu_tool.py

```diff
@@ -197,41 +225,31 @@ def slice_compilation_db(compilation_db, selection):
 
     return new_db
 
+def execute(invocations, verbose, formatter, jobs):
+    if jobs == 1:
+        for invocation in invocations:
+            print(formatter(invocation.run(verbose).get_output()))
+        return
 
-def _run_iwyu(invocation, verbose):
-    """ Start the process described by invocation.
-
-    Only necessary because multiprocessing can't pickle instance methods.
-    """
-    return invocation.run(verbose)
-
+    pending = []
+    while invocations or pending:
+        # get complete iwyu processes
+        complete = [proc for proc in pending if proc.poll() is not None]
 
-def execute(invocations, verbose, formatter, jobs):
-    """ Execute all invocations with at most `jobs` in parallel. """
-    try:
-        pool = multiprocessing.Pool(jobs)
-        # No actual results in `results`, it's only used for exception handling.
-        # Details here: https://stackoverflow.com/a/28660669.
-        results = []
+        for proc in complete:
+            # handle iwyu output
+            pending.remove(proc)
+            print(formatter(proc.get_output()))
+            proc.close()
 
-        for invocation in invocations:
-            results.append(
-                pool.apply_async(
-                    _run_iwyu, (invocation, verbose),
-                    callback=lambda x: print(formatter(x))))
-        pool.close()
-        pool.join()
-        for r in results:
-            r.get()
-    except OSError as why:
-        print('ERROR: Failed to launch include-what-you-use: %s' % why)
-        return 1
+        n = jobs - len(pending)
+        pending.extend(e.run(verbose) for e in invocations[:n])
```

Comment:
Call `Process.start` here instead, it will be more obvious that that returns a Process object.

Reply:
Oh, I see -- there's no seam for mocking if you do. Hmm. 

I'd like to see the following changes, then:

- Rename `run` -> `start`
- Rename `e` -> `i` for "invocation"